src/v/storage/index_state.cc

```diff
@@ -124,4 +128,90 @@ iobuf index_state::checksum_and_serialize() {
     return serde_compat::index_state_serde::encode(*this);
 }
 
+void index_state::serde_write(iobuf& out) const {
+    using serde::write;
+
+    iobuf tmp;
+    write(tmp, bitflags);
+    write(tmp, base_offset);
+    write(tmp, max_offset);
+    write(tmp, base_timestamp);
+    write(tmp, max_timestamp);
+    write(tmp, relative_offset_index.copy());
+    write(tmp, relative_time_index.copy());
+    write(tmp, position_index.copy());
+
+    crc::crc32c crc;
+    crc_extend_iobuf(crc, tmp);
+    const uint32_t tmp_crc = crc.value();
+
+    // data blob + crc
+    write(out, std::move(tmp));
+    write(out, tmp_crc);
+}
+
+void read_nested(
+  iobuf_parser& in, index_state& st, const size_t bytes_left_limit) {
+    /*
+     * peek at the 1-byte version prefix. serde support starts with the version
+     * immediately following the on-disk deprecated version.
+     */
+    if (
+      serde::peek_version(in)
+      > serde_compat::index_state_serde::ondisk_version) {
+        // consume the full serde header from the input
+        const auto hdr = serde::read_header<index_state>(in, bytes_left_limit);
+
+        using serde::read_nested;
+
+        // data blog + crc
+        iobuf tmp;
+        uint32_t tmp_crc = 0;
+        read_nested(in, tmp, hdr._bytes_left_limit);
+        read_nested(in, tmp_crc, hdr._bytes_left_limit);
+
+        crc::crc32c crc;
+        crc_extend_iobuf(crc, tmp);
+        const uint32_t expected_tmp_crc = crc.value();
+
+        if (tmp_crc != expected_tmp_crc) {
+            throw serde::serde_exception(fmt_with_ctx(
+              fmt::format,
+              "Mismatched checksum {} expected {}",
+              tmp_crc,
+              expected_tmp_crc));
+        }
+
+        // unwrap actual fields
+        iobuf_parser p(std::move(tmp));
+        st.bitflags = read_nested<decltype(st.bitflags)>(p, 0U);
+        st.base_offset = read_nested<decltype(st.base_offset)>(p, 0U);
+        st.max_offset = read_nested<decltype(st.max_offset)>(p, 0U);
+        st.base_timestamp = read_nested<decltype(st.base_timestamp)>(p, 0U);
+        st.max_timestamp = read_nested<decltype(st.max_timestamp)>(p, 0U);
+        st.relative_offset_index
+          = read_nested<decltype(st.relative_offset_index)>(p, 0U);
+        st.relative_time_index = read_nested<decltype(st.relative_time_index)>(
+          p, 0U);
+        st.position_index = read_nested<decltype(st.position_index)>(p, 0U);
```

Comment:
@felixguendling I only figured out how to do this by passing in the `0U` parameter as well as declaring the template type explicitly. I wonder if we can create an interface that removes that `0U` parameter and does type deduction on an l-value reference?

also, open to other ways of re-writing all of this in general. I'm realize there is a checksum-specific envelope but I think it is a good idea to write this out manually because in the general case we'll hit some other specific scenario that we need to handle manually.